From 6a61c8c08e4f06cf07a5e6b8be154d37007f8a45 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Micha=C5=82=20G=C3=B3rny?= <mgorny@gentoo.org>
Date: Thu, 21 Nov 2024 15:02:04 +0100
Subject: [PATCH 2/4] Do not link directly to LLVM static libraries

Remove direct linking to LLVM static libraries, use the shared library
enforced by MLIR instead.  It is incorrect to simultaneously link to
shared and static LLVM libraries, and it results in errors:

    : CommandLine Error: Option 'print-pipeline-passes' registered more than once!

Since static MLIR libraries expose an indirect dependency on the shared
LLVM library, remove direct linking to static libraries.
---
 CMakeLists.txt     | 27 ---------------------------
 bin/CMakeLists.txt |  6 ------
 2 files changed, 33 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index d8cadd8bf..d72559502 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -237,37 +237,10 @@ if(TRITON_BUILD_PYTHON_MODULE)
     MLIRGPUToROCDLTransforms
     MLIRUBToLLVM
 
-    # LLVM
-    LLVMPasses
-    LLVMNVPTXCodeGen
-    # LLVMNVPTXAsmPrinter
-    LLVMAMDGPUCodeGen
-    LLVMAMDGPUAsmParser
-
     Python3::Module
     pybind11::headers
 
   )
-  if(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64" OR # Linux arm64
-     CMAKE_SYSTEM_PROCESSOR MATCHES "arm64" OR # macOS arm64
-     CMAKE_OSX_ARCHITECTURES MATCHES "arm64")  # also macOS arm64
-      list(APPEND TRITON_LIBRARIES
-          LLVMAArch64CodeGen
-          LLVMAArch64AsmParser
-      )
-  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "x86_64" OR CMAKE_SYSTEM_PROCESSOR MATCHES "AMD64")
-      list(APPEND TRITON_LIBRARIES
-          LLVMX86CodeGen
-          LLVMX86AsmParser
-      )
-  elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "ppc64le")
-      list(APPEND TRITON_LIBRARIES
-        LLVMPowerPCAsmParser
-        LLVMPowerPCCodeGen
-      )
-  else()
-    message(FATAL_ERROR "LLVM codegen/ASM parser libs: This HW architecture (${CMAKE_SYSTEM_PROCESSOR}) is not configured in cmake lib dependencies.")
-  endif()
 
   # Define triton library
   string(JOIN "," TRITON_BACKENDS_TUPLE ${TRITON_CODEGEN_BACKENDS})
diff --git a/bin/CMakeLists.txt b/bin/CMakeLists.txt
index 7ed6d6270..cc1124fcc 100644
--- a/bin/CMakeLists.txt
+++ b/bin/CMakeLists.txt
@@ -72,12 +72,6 @@ add_llvm_executable(triton-llvm-opt
   )
 target_link_libraries(triton-llvm-opt PRIVATE
   TritonLLVMIR
-
-  LLVMAnalysis
-  LLVMCore
-  LLVMSupport
-  LLVMOption
-  LLVMCodeGen
   )
 export_executable_symbols_for_plugins(triton-llvm-opt)
 
